File: plugins/gemini-infographics/skills/gemini-infographics/scripts/generate-infographics.mjs
#!/usr/bin/env node
// @ts-check
/**
 * Generate infographics with the Gemini image API ("nano banana").
 *
 * Usage:
 *   generate-infographics.mjs SPEC.json --out DIR [--mode auto|sync|batch]
 *                             [--model MODEL] [--no-fallback] [--dry-run]
 *   generate-infographics.mjs --list-models
 *
 * SPEC.json is a list of entries:
 *   [{"name": "viz-x", "prompt": "...", "aspectRatio": "16:9",
 *     "imageSize": "1K", "model": "flash"}]
 *
 * Only `name` and `prompt` are required. Results are written to DIR/<name>.png.
 * Entries sharing a resolved model are batched (half price) when there are >= 3
 * of them and mode is auto; otherwise they run sync. Exit code is non-zero if
 * any image failed after retries.
 *
 * Model selection, highest precedence first:
 *   1. the entry's "model"
 *   2. --model
 *   3. $GEMINI_IMAGE_MODEL
 *   4. DEFAULT_ALIAS below
 *
 * A model may be a concrete id ("gemini-3.1-flash-image") or an alias:
 *   flash | flash-lite | pro    pinned ids, the known-good defaults
 *   latest                      newest image model the key can see
 *   latest-flash | latest-flash-lite | latest-pro
 *                               newest model in that family
 *
 * Pinned ids are resolved live too: if one is gone (404/NOT_FOUND) or the key
 * has no access to it, the script falls back to the newest model in the same
 * family and keeps going. `--no-fallback` (or GEMINI_IMAGE_MODEL_FALLBACK=off)
 * turns that off and lets the failure stand.
 *
 * Node >= 18 standard library only — no dependencies, no build step.
 */

import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { parseArgs } from "node:util";

/**
 * @typedef {object} Entry
 * @property {string} name
 * @property {string} prompt
 * @property {string} [aspectRatio]
 * @property {string} [imageSize]
 * @property {string} [model]
 */

const API_BASE =
  process.env.GEMINI_API_BASE ??
  "https://generativelanguage.googleapis.com/v1beta";
const BATCH_THRESHOLD = 3;
const BATCH_TIMEOUT_MS = 2_400_000;
const SYNC_RETRIES = 3;
const POLL_INTERVAL_MS = 15_000;

const DEFAULT_ALIAS = "flash";

/**
 * Pinned ids are what these aliases mean *today*; the resolver below repairs
 * them against the live model list when an id stops existing.
 * @type {Record<string, string>}
 */
const ALIASES = {
  flash: "gemini-3.1-flash-image",
  "flash-lite": "gemini-3.1-flash-lite-image",
  pro: "gemini-3-pro-image",
};

/** @type {Record<string, string>} */
const LATEST_ALIASES = {
  latest: "any",
  "latest-flash": "flash",
  "latest-flash-lite": "flash-lite",
  "latest-pro": "pro",
};

/** @type {string[]} */
let modelsCache = [];
/** @type {Map<string, string>} */
const latestCache = new Map();
/**
 * Concrete id -> replacement, once a fallback has been resolved for it. Keeps
 * a dead pin from costing one ListModels round trip per image.
 * @type {Map<string, string>}
 */
const remapped = new Map();

/** @param {string} msg @returns {never} */
function die(msg) {
  console.error(`ERROR: ${msg}`);
  process.exit(1);
}

/** @param {number} ms */
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

function apiKey() {
  const key = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
  if (!key) {
    die(
      "GEMINI_API_KEY is not set.\n" +
        "Set it with:  export GEMINI_API_KEY='your-key-here'\n" +
        "Get a key at: https://aistudio.google.com/apikey",
    );
  }
  return key;
}

/** An API response that came back with a non-2xx status. */
class HttpError extends Error {
  /** @param {number} status @param {string} detail */
  constructor(status, detail) {
    super(`HTTP ${status} ${detail}`);
    this.status = status;
    this.detail = detail;
  }
}

/**
 * @param {Response} res
 * @returns {Promise<any>}
 */
async function readJson(res) {
  if (!res.ok) {
    throw new HttpError(res.status, (await res.text()).slice(0, 200));
  }
  return res.json();
}

/**
 * @param {string} url
 * @param {unknown} body
 * @param {number} [timeoutMs]
 * @returns {Promise<any>}
 */
async function post(url, body, timeoutMs = 300_000) {
  return readJson(
    await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-goog-api-key": apiKey(),
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(timeoutMs),
    }),
  );
}

/**
 * @param {string} url
 * @param {number} [timeoutMs]
 * @returns {Promise<any>}
 */
async function get(url, timeoutMs = 120_000) {
  return readJson(
    await fetch(url, {
      headers: { "x-goog-api-key": apiKey() },
      signal: AbortSignal.timeout(timeoutMs),
    }),
  );
}

// ---------------------------------------------------------------------------
// model resolution

/** @param {string} model */
function familyOf(model) {
  if (model.includes("flash-lite")) return "flash-lite";
  if (model.includes("flash")) return "flash";
  if (model.includes("pro")) return "pro";
  return "any";
}

/** Sort key: version number first, stable before preview/exp. @param {string} name */
function versionKey(name) {
  const m = /gemini-(\d+)(?:\.(\d+))?/.exec(name);
  const major = m ? Number(m[1]) : 0;
  const minor = m && m[2] ? Number(m[2]) : 0;
  const stable = /preview|exp|-\d{2}-\d{4}$/.test(name) ? 0 : 1;
  return [major, minor, stable, -name.length];
}

/** @param {string} a @param {string} b */
function byNewestFirst(a, b) {
  const ka = versionKey(a);
  const kb = versionKey(b);
  for (let i = 0; i < ka.length; i++) {
    const diff = (kb[i] ?? 0) - (ka[i] ?? 0);
    if (diff !== 0) return diff;
  }
  return 0;
}

/** Every image-generating model the key can see, newest first. */
async function imageModels() {
  if (modelsCache.length) return modelsCache;
  /** @type {any[]} */
  const page = [];
  let token = "";
  for (;;) {
    const url = `${API_BASE}/models?pageSize=200${token ? `&pageToken=${token}` : ""}`;
    const data = await get(url);
    page.push(...(data.models ?? []));
    token = data.nextPageToken ?? "";
    if (!token) break;
  }
  /** @type {string[]} */
  const found = [];
  for (const m of page) {
    const name = String(m.name ?? "").split("/").pop() ?? "";
    // An image model is one whose name says so; the generation-method list is
    // "generateContent" for image and text models alike.
    const methods = m.supportedGenerationMethods ?? [];
    if (/-image(-|$)/.test(name) && methods.includes("generateContent")) {
      found.push(name);
    }
  }
  modelsCache = found.sort(byNewestFirst);
  return modelsCache;
}

/** @param {string} family */
async function latestModel(family) {
  const cached = latestCache.get(family);
  if (cached) return cached;
  const candidates = (await imageModels()).filter(
    (m) => family === "any" || familyOf(m) === family,
  );
  const newest = candidates[0];
  if (!newest) {
    die(
      `no image model available for family '${family}'.\n` +
        "Run with --list-models to see what this API key can reach, then " +
        "pin one explicitly with --model.",
    );
  }
  latestCache.set(family, newest);
  return newest;
}

/**
 * Alias or concrete id -> concrete id. Only 'latest*' hits the network.
 * @param {string} raw
 */
async function resolveModel(raw) {
  if (raw in ALIASES) return /** @type {string} */ (ALIASES[raw]);
  const family = LATEST_ALIASES[raw];
  if (family) return latestModel(family);
  return raw;
}

/** @param {unknown} err */
function isMissingModel(err) {
  if (!(err instanceof HttpError)) return false;
  return (
    (err.status === 403 || err.status === 404) &&
    /NOT_FOUND|not found|PERMISSION_DENIED|does not exist/.test(err.detail)
  );
}

/**
 * Newest live model in `model`'s family, or `model` itself if none / off.
 * @param {string} model
 * @param {boolean} fallback
 */
async function fallbackModel(model, fallback) {
  if (!fallback) return model;
  const known = remapped.get(model);
  if (known) return known;
  const replacement = await latestModel(familyOf(model));
  if (replacement !== model) {
    console.error(`  ${model} unavailable — falling back to ${replacement}`);
  }
  remapped.set(model, replacement);
  return replacement;
}

// ---------------------------------------------------------------------------
// generation

/** @param {Entry} entry */
function genRequest(entry) {
  return {
    contents: [{ parts: [{ text: entry.prompt }] }],
    generationConfig: {
      responseModalities: ["TEXT", "IMAGE"],
      imageConfig: {
        aspectRatio: entry.aspectRatio ?? "16:9",
        imageSize: entry.imageSize ?? "1K",
      },
    },
  };
}

/** @param {any} response @param {string} path */
function saveImage(response, path) {
  for (const cand of response?.candidates ?? []) {
    for (const part of cand?.content?.parts ?? []) {
      if (part.inlineData) {
        writeFileSync(path, Buffer.from(part.inlineData.data, "base64"));
        return true;
      }
    }
  }
  return false;
}

/**
 * @param {string} model
 * @param {Entry[]} entries
 * @param {string} outdir
 * @param {boolean} fallback
 * @returns {Promise<string[]>} names that failed
 */
async function runSync(model, entries, outdir, fallback) {
  /** @type {string[]} */
  const failed = [];
  for (const e of entries) {
    // a fallback resolved for an earlier entry applies to the rest of the
    // group — never re-discover a dead pin per image
    model = remapped.get(model) ?? model;
    const path = join(outdir, `${e.name}.png`);
    let ok = false;
    for (let attempt = 1; attempt <= SYNC_RETRIES; attempt++) {
      /** @type {any} */
      let resp;
      try {
        resp = await post(
          `${API_BASE}/models/${model}:generateContent`,
          genRequest(e),
        );
      } catch (err) {
        if (err instanceof HttpError) {
          console.error(`  ${e.name}: HTTP ${err.status} ${err.detail}`);
          if (isMissingModel(err)) {
            const replacement = await fallbackModel(model, fallback);
            if (replacement !== model) {
              model = replacement;
              continue;
            }
            break;
          }
          if (
            [429, 500, 503].includes(err.status) &&
            attempt < SYNC_RETRIES
          ) {
            await sleep(15_000 * attempt);
            continue;
          }
          break;
        }
        console.error(`  ${e.name}: ${String(err)}`);
        if (attempt === SYNC_RETRIES) break;
        await sleep(5_000);
        continue;
      }
      if (saveImage(resp, path)) {
        console.log(`  ok ${e.name}`);
        ok = true;
        break;
      }
      console.error(`  ${e.name}: empty response, attempt ${attempt}`);
      await sleep(5_000);
    }
    if (!ok) failed.push(e.name);
  }
  return failed;
}

/**
 * @param {string} model
 * @param {Entry[]} entries
 * @param {string} outdir
 * @param {boolean} fallback
 * @returns {Promise<string[]>} names that failed
 */
async function runBatch(model, entries, outdir, fallback) {
  const body = {
    batch: {
      display_name: `infographics-${Date.now()}`,
      input_config: {
        requests: {
          requests: entries.map((e) => ({
            request: genRequest(e),
            metadata: { key: e.name },
          })),
        },
      },
    },
  };

  /** @type {any} */
  let op;
  try {
    op = await post(
      `${API_BASE}/models/${model}:batchGenerateContent`,
      body,
      120_000,
    );
  } catch (err) {
    console.error(
      `  batch submit: ${err instanceof HttpError ? `HTTP ${err.status} ${err.detail}` : String(err)}`,
    );
    if (isMissingModel(err)) {
      const replacement = await fallbackModel(model, fallback);
      if (replacement !== model) {
        return runBatch(replacement, entries, outdir, fallback);
      }
    }
    console.error("  batch submit failed — falling back to sync");
    return runSync(model, entries, outdir, fallback);
  }

  const opname = op.name;
  console.log(`  batch ${opname} (${entries.length} images)`);

  const t0 = Date.now();
  /** @type {any} */
  let st;
  for (;;) {
    await sleep(POLL_INTERVAL_MS);
    // transient network errors must not abandon a paid batch — keep polling
    try {
      st = await get(`${API_BASE}/${opname}`);
    } catch (err) {
      console.error(
        `  ${elapsed(t0)} poll error (${String(err)}); retrying`,
      );
      continue;
    }
    console.log(`  ${elapsed(t0)} ${st?.metadata?.state ?? "?"}`);
    if (st.done) break;
    if (Date.now() - t0 > BATCH_TIMEOUT_MS) {
      console.error("  batch timeout — falling back to sync");
      return runSync(model, entries, outdir, fallback);
    }
  }

  const byName = new Map(entries.map((e) => [e.name, e]));
  /** @type {string[]} */
  const failed = [];
  const inlined = st?.response?.inlinedResponses?.inlinedResponses ?? [];
  for (const item of inlined) {
    const key = item?.metadata?.key ?? "?";
    if (item.error) {
      console.error(`  ERR ${key}: ${JSON.stringify(item.error).slice(0, 150)}`);
      failed.push(key);
      continue;
    }
    if (saveImage(item.response, join(outdir, `${key}.png`))) {
      console.log(`  ok ${key}`);
    } else {
      console.error(`  ${key}: no image in batch response`);
      failed.push(key);
    }
  }

  // per-request failures inside a successful batch retry individually
  const retry = failed
    .map((n) => byName.get(n))
    .filter(/** @returns {e is Entry} */ (e) => e !== undefined);
  if (retry.length) {
    console.log(`  retrying ${retry.length} failed entries sync`);
    return runSync(model, retry, outdir, fallback);
  }
  return [];
}

/** @param {number} t0 */
function elapsed(t0) {
  return `${String(Math.round((Date.now() - t0) / 1000)).padStart(5)}s`;
}

// ---------------------------------------------------------------------------

const USAGE = `Usage:
  generate-infographics.mjs SPEC.json --out DIR [--mode auto|sync|batch]
                            [--model MODEL] [--no-fallback] [--dry-run]
  generate-infographics.mjs --list-models

  --out DIR        output directory for PNGs
  --mode MODE      auto (default) | sync | batch
  --model MODEL    default model for entries with no "model" (alias or id);
                   overrides $GEMINI_IMAGE_MODEL
  --no-fallback    fail instead of falling back to the latest model in the family
  --dry-run        validate the spec and print the model plan; generate nothing
  --list-models    list the image models this API key can reach, newest first
`;

/** @param {string} specPath @returns {Entry[]} */
function loadSpec(specPath) {
  /** @type {unknown} */
  let parsed;
  try {
    parsed = JSON.parse(readFileSync(specPath, "utf8"));
  } catch (err) {
    return die(`cannot read spec ${specPath}: ${String(err)}`);
  }
  if (!Array.isArray(parsed) || parsed.length === 0) {
    return die("spec must be a non-empty JSON list");
  }
  /** @type {Set<string>} */
  const seen = new Set();
  for (const e of parsed) {
    if (!e || typeof e.name !== "string" || typeof e.prompt !== "string") {
      die(`spec entry missing name/prompt: ${JSON.stringify(e).slice(0, 100)}`);
    }
    if (seen.has(e.name)) {
      die(`duplicate entry name '${e.name}' — outputs would overwrite`);
    }
    seen.add(e.name);
  }
  return parsed;
}

async function main() {
  const { values, positionals } = parseArgs({
    allowPositionals: true,
    options: {
      out: { type: "string" },
      mode: { type: "string", default: "auto" },
      model: { type: "string" },
      "no-fallback": { type: "boolean", default: false },
      "dry-run": { type: "boolean", default: false },
      "list-models": { type: "boolean", default: false },
      help: { type: "boolean", default: false, short: "h" },
    },
  });

  if (values.help) {
    console.log(USAGE);
    return;
  }

  if (values["list-models"]) {
    for (const m of await imageModels()) console.log(m);
    return;
  }

  const spec = positionals[0];
  const out = values.out;
  if (!spec || !out) {
    console.error(USAGE);
    die("spec and --out are required (unless --list-models)");
    return;
  }
  const mode = values.mode ?? "auto";
  if (!["auto", "sync", "batch"].includes(mode)) {
    die(`--mode must be auto, sync or batch (got '${mode}')`);
  }

  const fallback =
    !values["no-fallback"] && process.env.GEMINI_IMAGE_MODEL_FALLBACK !== "off";
  const defaultModel =
    values.model || process.env.GEMINI_IMAGE_MODEL || DEFAULT_ALIAS;

  const entries = loadSpec(spec);

  if (values["dry-run"]) {
    console.log(
      `default model: ${defaultModel}   fallback: ${fallback ? "on" : "off"}`,
    );
    for (const e of entries) {
      const raw = e.model || defaultModel;
      const shown = raw in LATEST_ALIASES ? raw : await resolveModel(raw);
      console.log(
        `  ${e.name}: ${raw} -> ${shown}` +
          `  ${e.aspectRatio ?? "16:9"} ${e.imageSize ?? "1K"}`,
      );
    }
    return;
  }

  mkdirSync(out, { recursive: true });
  apiKey();

  /** @type {Map<string, Entry[]>} */
  const byModel = new Map();
  for (const e of entries) {
    const model = await resolveModel(e.model || defaultModel);
    byModel.set(model, [...(byModel.get(model) ?? []), e]);
  }

  /** @type {string[]} */
  let failed = [];
  const t0 = Date.now();
  for (const [model, group] of byModel) {
    const useBatch =
      mode === "batch" || (mode === "auto" && group.length >= BATCH_THRESHOLD);
    console.log(
      `${model}: ${group.length} images (${useBatch ? "batch" : "sync"})`,
    );
    failed = failed.concat(
      useBatch
        ? await runBatch(model, group, out, fallback)
        : await runSync(model, group, out, fallback),
    );
  }

  console.log(
    `done: ${entries.length - failed.length}/${entries.length} in ` +
      `${Math.round((Date.now() - t0) / 1000)}s`,
  );
  if (failed.length) {
    console.error(`FAILED: ${failed.join(", ")}`);
    process.exit(2);
  }
}

await main();

File: plugins/gemini-infographics/skills/gemini-infographics/scripts/generate-infographics.py
#!/usr/bin/env python3
"""Generate infographics with the Gemini image API ("nano banana").

Usage:
  generate-infographics.py SPEC.json --out DIR [--mode auto|sync|batch]
                           [--model MODEL] [--no-fallback] [--dry-run]
  generate-infographics.py --list-models

SPEC.json is a list of entries:
  [{"name": "viz-x", "prompt": "...", "aspectRatio": "16:9",
    "imageSize": "1K", "model": "flash"}]

Only `name` and `prompt` are required. Results are written to DIR/<name>.png.
Entries sharing a resolved model are batched (half price) when there are >= 3
of them and mode is auto; otherwise they run sync. Exit code is non-zero if
any image failed after retries.

Model selection, highest precedence first:
  1. the entry's "model"
  2. --model
  3. $GEMINI_IMAGE_MODEL
  4. DEFAULT_ALIAS below

A model may be a concrete id ("gemini-3.1-flash-image") or an alias:
  flash | flash-lite | pro    pinned ids, the known-good defaults
  latest                      newest image model the key can see
  latest-flash | latest-flash-lite | latest-pro
                              newest model in that family

Pinned ids are resolved live too: if one is gone (404/NOT_FOUND) or the key
has no access to it, the script falls back to the newest model in the same
family and keeps going. `--no-fallback` (or GEMINI_IMAGE_MODEL_FALLBACK=off)
turns that off and lets the failure stand.
"""

import argparse
import base64
import json
import os
import re
import sys
import time
import urllib.error
import urllib.request

API_BASE = os.environ.get(
    "GEMINI_API_BASE", "https://generativelanguage.googleapis.com/v1beta"
)
BATCH_THRESHOLD = 3
BATCH_TIMEOUT_S = 2400
SYNC_RETRIES = 3

DEFAULT_ALIAS = "flash"

# Pinned ids are what these aliases mean *today*; the resolver below repairs
# them against the live model list when an id stops existing.
ALIASES = {
    "flash": ("gemini-3.1-flash-image", "flash"),
    "flash-lite": ("gemini-3.1-flash-lite-image", "flash-lite"),
    "pro": ("gemini-3-pro-image", "pro"),
}
LATEST_ALIASES = {
    "latest": "any",
    "latest-flash": "flash",
    "latest-flash-lite": "flash-lite",
    "latest-pro": "pro",
}

_models_cache: list = []
_latest_cache: dict = {}
# Concrete id -> replacement, once a fallback has been resolved for it. Keeps
# a dead pin from costing one ListModels round trip per entry.
_remapped: dict = {}


def die(msg: str) -> None:
    print(f"ERROR: {msg}", file=sys.stderr)
    sys.exit(1)


def api_key() -> str:
    key = os.environ.get("GEMINI_API_KEY", "") or os.environ.get("GOOGLE_API_KEY", "")
    if not key:
        die(
            "GEMINI_API_KEY is not set.\n"
            "Set it with:  export GEMINI_API_KEY='your-key-here'\n"
            "Get a key at: https://aistudio.google.com/apikey"
        )
    return key


def post(url: str, body: dict, timeout: int = 300) -> dict:
    req = urllib.request.Request(
        url,
        data=json.dumps(body).encode(),
        headers={"Content-Type": "application/json", "x-goog-api-key": api_key()},
    )
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.load(r)


def get(url: str, timeout: int = 120) -> dict:
    req = urllib.request.Request(url, headers={"x-goog-api-key": api_key()})
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.load(r)


# --------------------------------------------------------------------------
# model resolution


def family_of(model: str) -> str:
    if "flash-lite" in model:
        return "flash-lite"
    if "flash" in model:
        return "flash"
    if "pro" in model:
        return "pro"
    return "any"


def image_models() -> list:
    """Every image-generating model the key can see, newest first."""
    if _models_cache:
        return _models_cache
    page, token = [], None
    while True:
        url = f"{API_BASE}/models?pageSize=200" + (f"&pageToken={token}" if token else "")
        data = get(url)
        page += data.get("models", [])
        token = data.get("nextPageToken")
        if not token:
            break
    for m in page:
        name = m.get("name", "").split("/")[-1]
        methods = m.get("supportedGenerationMethods", [])
        # An image model is one whose name says so; the generation-method list
        # is "generateContent" for image and text models alike.
        if "generateContent" in methods and re.search(r"-image(-|$)", name):
            _models_cache.append(name)
    _models_cache.sort(key=version_key, reverse=True)
    return _models_cache


def version_key(name: str) -> tuple:
    """Sort key: version number first, stable before preview/exp."""
    m = re.search(r"gemini-(\d+)(?:\.(\d+))?", name)
    major, minor = (int(m.group(1)), int(m.group(2) or 0)) if m else (0, 0)
    stable = 0 if re.search(r"preview|exp|-\d{2}-\d{4}$", name) else 1
    return (major, minor, stable, -len(name))


def latest_model(family: str) -> str:
    if family in _latest_cache:
        return _latest_cache[family]
    candidates = [m for m in image_models() if family == "any" or family_of(m) == family]
    if not candidates:
        die(
            f"no image model available for family '{family}'.\n"
            "Run with --list-models to see what this API key can reach, then "
            "pin one explicitly with --model."
        )
    _latest_cache[family] = candidates[0]
    return candidates[0]


def resolve_model(raw: str) -> str:
    """Alias or concrete id -> concrete id. Only 'latest*' hits the network."""
    if raw in ALIASES:
        return ALIASES[raw][0]
    if raw in LATEST_ALIASES:
        return latest_model(LATEST_ALIASES[raw])
    return raw


def is_missing_model(err: urllib.error.HTTPError, detail: str) -> bool:
    return err.code in (403, 404) and (
        "NOT_FOUND" in detail
        or "not found" in detail
        or "PERMISSION_DENIED" in detail
        or "does not exist" in detail
    )


def fallback_model(model: str, fallback: bool) -> str:
    """Newest live model in `model`'s family, or `model` itself if none/off."""
    if not fallback:
        return model
    if model in _remapped:
        return _remapped[model]
    try:
        replacement = latest_model(family_of(model))
    except SystemExit:
        return model
    if replacement != model:
        print(
            f"  {model} unavailable — falling back to {replacement}", file=sys.stderr
        )
    _remapped[model] = replacement
    return replacement


# --------------------------------------------------------------------------
# generation


def gen_request(entry: dict) -> dict:
    cfg = {
        "responseModalities": ["TEXT", "IMAGE"],
        "imageConfig": {
            "aspectRatio": entry.get("aspectRatio", "16:9"),
            "imageSize": entry.get("imageSize", "1K"),
        },
    }
    return {
        "contents": [{"parts": [{"text": entry["prompt"]}]}],
        "generationConfig": cfg,
    }


def save_image(response: dict, path: str) -> bool:
    for cand in response.get("candidates", []):
        for part in cand.get("content", {}).get("parts", []):
            if "inlineData" in part:
                with open(path, "wb") as f:
                    f.write(base64.b64decode(part["inlineData"]["data"]))
                return True
    return False


def run_sync(model: str, entries: list, outdir: str, fallback: bool) -> list:
    failed = []
    for e in entries:
        # a fallback resolved for an earlier entry applies to the rest of the
        # group — never re-discover a dead pin per image
        model = _remapped.get(model, model)
        path = os.path.join(outdir, f"{e['name']}.png")
        ok = False
        for attempt in range(1, SYNC_RETRIES + 1):
            try:
                resp = post(f"{API_BASE}/models/{model}:generateContent", gen_request(e))
            except urllib.error.HTTPError as err:
                detail = err.read().decode()[:200]
                print(f"  {e['name']}: HTTP {err.code} {detail}", file=sys.stderr)
                if is_missing_model(err, detail):
                    replacement = fallback_model(model, fallback)
                    if replacement != model:
                        model = replacement
                        continue
                    break
                if err.code in (429, 500, 503) and attempt < SYNC_RETRIES:
                    time.sleep(15 * attempt)
                    continue
                break
            if save_image(resp, path):
                print(f"  ok {e['name']}")
                ok = True
                break
            print(f"  {e['name']}: empty response, attempt {attempt}", file=sys.stderr)
            time.sleep(5)
        if not ok:
            failed.append(e["name"])
    return failed


def run_batch(model: str, entries: list, outdir: str, fallback: bool) -> list:
    reqs = [
        {"request": gen_request(e), "metadata": {"key": e["name"]}} for e in entries
    ]
    body = {
        "batch": {
            "display_name": f"infographics-{int(time.time())}",
            "input_config": {"requests": {"requests": reqs}},
        }
    }
    try:
        op = post(f"{API_BASE}/models/{model}:batchGenerateContent", body, timeout=120)
    except urllib.error.HTTPError as err:
        detail = err.read().decode()[:200]
        print(f"  batch submit: HTTP {err.code} {detail}", file=sys.stderr)
        if is_missing_model(err, detail):
            replacement = fallback_model(model, fallback)
            if replacement != model:
                return run_batch(replacement, entries, outdir, fallback)
        print("  batch submit failed — falling back to sync", file=sys.stderr)
        return run_sync(model, entries, outdir, fallback)
    opname = op["name"]
    print(f"  batch {opname} ({len(entries)} images)")

    t0 = time.time()
    while True:
        time.sleep(15)
        # transient network errors must not abandon a paid batch — keep polling
        try:
            st = get(f"{API_BASE}/{opname}")
        except (urllib.error.URLError, OSError) as err:
            print(
                f"  {time.time() - t0:5.0f}s poll error ({err}); retrying",
                file=sys.stderr,
            )
            continue
        state = st.get("metadata", {}).get("state", "?")
        print(f"  {time.time() - t0:5.0f}s {state}")
        if st.get("done"):
            break
        if time.time() - t0 > BATCH_TIMEOUT_S:
            print("  batch timeout — falling back to sync", file=sys.stderr)
            return run_sync(model, entries, outdir, fallback)

    by_name = {e["name"]: e for e in entries}
    failed = []
    inlined = (
        st.get("response", {}).get("inlinedResponses", {}).get("inlinedResponses", [])
    )
    for item in inlined:
        key = item.get("metadata", {}).get("key", "?")
        if "error" in item:
            print(f"  ERR {key}: {json.dumps(item['error'])[:150]}", file=sys.stderr)
            failed.append(key)
            continue
        path = os.path.join(outdir, f"{key}.png")
        if save_image(item.get("response", {}), path):
            print(f"  ok {key}")
        else:
            print(f"  {key}: no image in batch response", file=sys.stderr)
            failed.append(key)

    # per-request failures inside a successful batch retry individually
    retry = [by_name[n] for n in failed if n in by_name]
    if retry:
        print(f"  retrying {len(retry)} failed entries sync")
        return run_sync(model, retry, outdir, fallback)
    return []


# --------------------------------------------------------------------------


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("spec", nargs="?", help="path to spec JSON")
    ap.add_argument("--out", help="output directory for PNGs")
    ap.add_argument("--mode", choices=["auto", "sync", "batch"], default="auto")
    ap.add_argument(
        "--model",
        help="default model for entries with no 'model' (alias or id); "
        "overrides $GEMINI_IMAGE_MODEL",
    )
    ap.add_argument(
        "--no-fallback",
        action="store_true",
        help="fail instead of falling back to the latest model in the family",
    )
    ap.add_argument(
        "--dry-run",
        action="store_true",
        help="validate the spec and print the model plan; generate nothing",
    )
    ap.add_argument(
        "--list-models",
        action="store_true",
        help="list the image models this API key can reach, newest first",
    )
    args = ap.parse_args()

    if args.list_models:
        for m in image_models():
            print(m)
        return

    if not args.spec or not args.out:
        ap.error("spec and --out are required (unless --list-models)")

    fallback = not (
        args.no_fallback or os.environ.get("GEMINI_IMAGE_MODEL_FALLBACK") == "off"
    )
    default_model = args.model or os.environ.get("GEMINI_IMAGE_MODEL") or DEFAULT_ALIAS

    with open(args.spec) as f:
        entries = json.load(f)
    if not isinstance(entries, list) or not entries:
        die("spec must be a non-empty JSON list")
    seen = set()
    for e in entries:
        if "name" not in e or "prompt" not in e:
            die(f"spec entry missing name/prompt: {json.dumps(e)[:100]}")
        if e["name"] in seen:
            die(f"duplicate entry name '{e['name']}' — outputs would overwrite")
        seen.add(e["name"])

    if args.dry_run:
        print(f"default model: {default_model}   fallback: {'on' if fallback else 'off'}")
        for e in entries:
            raw = e.get("model") or default_model
            shown = raw if raw in LATEST_ALIASES else resolve_model(raw)
            print(
                f"  {e['name']}: {raw} -> {shown}"
                f"  {e.get('aspectRatio', '16:9')} {e.get('imageSize', '1K')}"
            )
        return

    os.makedirs(args.out, exist_ok=True)
    api_key()

    by_model: dict = {}
    for e in entries:
        by_model.setdefault(resolve_model(e.get("model") or default_model), []).append(e)

    failed: list = []
    t0 = time.time()
    for model, group in by_model.items():
        use_batch = args.mode == "batch" or (
            args.mode == "auto" and len(group) >= BATCH_THRESHOLD
        )
        print(f"{model}: {len(group)} images ({'batch' if use_batch else 'sync'})")
        if use_batch:
            failed += run_batch(model, group, args.out, fallback)
        else:
            failed += run_sync(model, group, args.out, fallback)

    print(
        f"done: {len(entries) - len(failed)}/{len(entries)} in {time.time() - t0:.0f}s"
    )
    if failed:
        print(f"FAILED: {', '.join(failed)}", file=sys.stderr)
        sys.exit(2)


if __name__ == "__main__":
    main()
